pkg/config/resource.go

```diff
@@ -20,40 +20,59 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 )
 
-// SetIdentifierArgumentFn sets the name of the resource in Terraform attributes map.
-type SetIdentifierArgumentFn func(base map[string]interface{}, name string)
+// SetIdentifierArgumentsFn sets the name of the resource in Terraform attributes map,
+// i.e. Main HCL file.
+type SetIdentifierArgumentsFn func(base map[string]interface{}, name string)
 
 // NopSetIdentifierArgument does nothing. It's useful for cases where the external
 // name is calculated by provider and doesn't have any effect on spec fields.
-func NopSetIdentifierArgument(_ map[string]interface{}, _ string) {}
+var NopSetIdentifierArgument SetIdentifierArgumentsFn = func(_ map[string]interface{}, _ string) {}
+
+// GetIDFn returns the ID to be used in TF State file, i.e. "id" field in
+// terraform.tfstate.
+type GetIDFn func(name string, parameters map[string]interface{}, providerConfig map[string]interface{}) string
+
+// NameAsID returns the name to be used as ID in TF State file.
+var NameAsID GetIDFn = func(name string, _ map[string]interface{}, _ map[string]interface{}) string {
+	return name
+}
+
+// GetNameFn returns the external name extracted from the TF State.
+type GetNameFn func(tfstate map[string]interface{}) string
+
+// IDAsName returns the TF State ID as external name.
+var IDAsName GetNameFn = func(tfstate map[string]interface{}) string {
+	if id, ok := tfstate["id"].(string); ok {
+		return id
+	}
+	return ""
+}
 
 // AdditionalConnectionDetailsFn functions adds custom keys to connection details
 // secret using input terraform attributes
 type AdditionalConnectionDetailsFn func(attr map[string]interface{}) (map[string][]byte, error)
 
 // NopAdditionalConnectionDetails does nothing, when no additional connection
 // details configuration function provided.
-func NopAdditionalConnectionDetails(_ map[string]interface{}) (map[string][]byte, error) {
+var NopAdditionalConnectionDetails AdditionalConnectionDetailsFn = func(_ map[string]interface{}) (map[string][]byte, error) {
 	return nil, nil
 }
 
-// ResourceOption allows setting optional fields of a Resource object.
-type ResourceOption func(*Resource)
-
-// WithTerraformIDFieldName allows you to set IDFieldName.
-func WithTerraformIDFieldName(n string) ResourceOption {
-	return func(c *Resource) {
-		c.IDFieldName = n
-	}
-}
-
 // ExternalName contains all information that is necessary for naming operations,
 // such as removal of those fields from spec schema and calling Configure function
 // to fill attributes with information given in external name.
 type ExternalName struct {
 	// SetIdentifierArgumentFn sets the name of the resource in Terraform argument
-	// map.
-	SetIdentifierArgumentFn SetIdentifierArgumentFn
+	// map, otherwise we cannot know which field in HCL we should treat as identifier.
```

Comment:
We had a similar discussion with @turkenh and my main reasons for explicitly relying on existence of these functions are mainly:
* If we assume no-op when it's not specified, we're making a decision on behalf of the developers and if it covers enough ground for initial tests, then they'll be aware of that decision later on rather than immediately.
* We have `DefaultResource` that makes this decision in a single place and it covers all those functions and also provide ways to override that provider-wide default so that they can very easily surface how that decision is made instead of diving into code where these functions are called.

So I'd prefer to keep requiring existence of this function. We can possibly have a validation somewhere in code-generation to check if they are `nil` so that we panic earlier.